toolchain/gcc: cleanup options and disable zstd explicitly

neheb opened the pull request:
Currently, toolchain/gcc picks up host zstd as it seems to be available in every Linux distro by default, even the minimal ones.

This PR as is has been tested to work perfectly fine on Debian 10 and Fedora 37. It has also been tested to completely break on Ubuntu 22.04 and 20.04.

As I can't seem to figure out what's going on with Ubuntu, posting this here.

edit: let's add more detail.

toolchain/gcc + --with-zstd=$(TOPDIR)/staging_dir/host + shared libzstd == rpath linking issue.
toolchain/gcc + --with-zstd=$(TOPDIR)/staging_dir/host + static libzstd == works except with Ubuntu
toolchain/gcc + system libzstd == works

Ubuntu error is that a zstd API call is failing during LTO. Two possibilities I can see.

1: Ubuntu's gcc has some out of tree patch that causes libzstd to be miscompiled and misbehave.
2: zstd source code has some Ubuntu specific workaround that needs to be removed.

I can't think of anything else.

edit2: specifically the error is during decompression, which indicates that either something is compressed wrong or the decompression routine is failing.

neheb committed:
tools/zstd: build libraries as static

Enables to get rid of rpath hack for all users.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
tools/cmake: remove rpath ldflag

no longer needed because of tools/zstd

Signed-off-by: Rosen Penev <rosenp@gmail.com>

Ansuel commented:
@mpratt14 want to have some fun with this?

mpratt14 commented:
yeah I already did, github instances have no problem with it...

so investigating the difference between github runner Ubuntu and native Ubuntu machines might reveal something

I dont have an ubuntu system to work on at the moment...

neheb commented:
If you have a Linux system might be worthwhile to run an nspaqn container.

mpratt14 commented:
before doing a deep dive into ubuntu, it may be helpful to play with the rest of the configuration

for example legacy_level, I caught this from the manual

```
  Limitation 2 : static cctx currently not compatible with multi-threading.
  Limitation 3 : static dctx is incompatible with legacy support.
```

setting legacy_level to 0 disables legacy support

```
if legacy_level == 0
  message('Legacy support: DISABLED')
else
  # See ZSTD_LEGACY_SUPPORT of lib/README.md
  message('Enable legacy support back to version 0.@0@'.format(legacy_level))
...
...
```

neheb commented:
Legacy level manipulation is one thing I tried. Didn't work for me.

mpratt14 commented:
instead of forcing multi-threading to work with the patch, what if we disable it?

neheb commented:
right. I mentioned that it makes one of the commits here obsolete.

mpratt14 commented:
so what youre saying is, it will work without multithreading, but that defeats the purpose which is saving time?

neheb commented:
No, disabling multithreading gets rid of the ptheead patch. Still broken on Ubuntu.

mpratt14 commented:
ok at this point, I actually have an ubuntu system to play with (updated an old laptop from ubuntu 16) and I cannot reproduce any issues. checking out this PR and a test branch which does git revert on the reverts with `pthread` patch over that, and no problems

```
Linux 5.15.0-48-generic #54-Ubuntu SMP Fri Aug 26 13:26:29 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

LSB Version:    core-11.1.0ubuntu4-noarch:security-11.1.0ubuntu4-noarch
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.1 LTS
Release:        22.04
Codename:       jammy
```

process of build:

```
git clone -b gcc https://github.com/neheb/openwrt.git gcc
```
OR
```
git clone -b ubuntu-zstd-test https://github.com/mpratt14/openwrt.git ubuntu-zstd-test
```
```
cd <topdir>
make package/symlinks
make defconfig; make prereq
nohup time ionice -c 3 make -j 4 -l 3 world &
```

Ansuel commented:
@mpratt14 same here we need consistent info on how to repro this...
@anomeome can you give us more details?

mpratt14 commented:
@anomeome can you test by actually checking out this branch exactly as it is? just so we are sure there's nothing else affecting the build results

for new buildroot:

```
git clone -b gcc https://github.com/neheb/openwrt.git
```

for existing buildroot:

```
git remote add neheb https://github.com/neheb/openwrt.git
git fetch neheb gcc
git checkout neheb/gcc
```

optionally make the branch local (instead of "detached HEAD" state)

```
git checkout -b <new-branch-name>
```

and do a full clean before building

```
make dirclean
```

mpratt14 commented:
I will try to build for mvebu_cortexa9 and see if that makes any difference...

mpratt14 commented:
it builds ok for me with your config except for using llvm and clang 14, but I think that's not related

```
< CONFIG_BPF_TOOLCHAIN_HOST_PATH="/usr/lib/llvm-15"
---
> CONFIG_BPF_TOOLCHAIN_HOST_PATH="/usr/lib/llvm-14"
```

Ansuel commented:
@mpratt14 it's either there was a broken package on ubuntu that got fixed or we are missing something really big here...

neheb commented:
@anomeome one way to see what's wrong is something like

```
find build_dir/toolchain_xxx -name config.log -exec cp --parents '{}' broken/working \;
```

Or something like that for broken and working setups.

neheb commented:
I still see this issue. Building only the gcc branch makes no difference

neheb commented:
Rebased.

Ansuel commented:
@neheb i'm compiling gcc for mipsel 24kc and 
```
OpenWrt-libtool: link: ccache_cc -shared  -fPIC -DPIC  .libs/rule.o .libs/statement.o .libs/cache.o .libs/cmd.o .libs/datatype.o .libs/expression.o .libs/evaluate.o .libs/proto.o .libs/payload.o .libs/exthdr.o .libs/fib.o .libs/hash.o .libs/intervals.o .libs/ipopt.o .libs/meta.o .libs/rt.o .libs/numgen.o .libs/ct.o .libs/xfrm.o .libs/netlink.o .libs/netlink_linearize.o .libs/netlink_delinearize.o .libs/misspell.o .libs/monitor.o .libs/owner.o .libs/segtree.o .libs/gmputil.o .libs/utils.o .libs/erec.o .libs/mnl.o .libs/iface.o .libs/mergesort.o .libs/optimize.o .libs/osf.o .libs/nfnl_osf.o .libs/tcpopt.o .libs/socket.o .libs/print.o .libs/sctp_chunk.o .libs/libnftables.o .libs/xt.o .libs/json.o .libs/parser_json.o  -Wl,--whole-archive ./.libs/libparser.a ./.libs/libminigmp.a -Wl,--no-whole-archive  -L/home/ansuel/openwrt/staging_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/usr/lib -L/home/ansuel/openwrt/staging_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/lib -L/home/ansuel/openwrt/staging_dir/target-mipsel_24kc_musl/usr/lib -lmnl -lnftnl -ljansson  -Os -mno-branch-likely -mips32r2 -mtune=24kc -msoft-float -mips16 -minterlink-mips16 -fstack-protector -Wl,-z -Wl,now -Wl,-z -Wl,relro -flto -Wl,--version-script=./libnftables.map -flto   -Wl,-soname -Wl,libnftables.so.1 -o .libs/libnftables.so.1.1.0
lto-wrapper: warning: using serial compilation of 9 LTRANS jobs
lto-wrapper: note: see the '-flto' option documentation for more information
lto1: internal compiler error: original not compressed with zstd
0x9417ba lto_uncompression_zstd
        /home/ansuel/openwrt/build_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/gcc-12.2.0/gcc/lto-compress.cc:163
0x9417ba lto_end_uncompression(lto_compression_stream*, lto_compression)
        /home/ansuel/openwrt/build_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/gcc-12.2.0/gcc/lto-compress.cc:406
0x93fe80 lto_get_section_data(lto_file_decl_data*, lto_section_type, char const*, int, unsigned long*, bool)
        /home/ansuel/openwrt/build_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/gcc-12.2.0/gcc/lto-section-in.cc:168
0x66906c cgraph_node::get_untransformed_body()
        /home/ansuel/openwrt/build_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/gcc-12.2.0/gcc/cgraph.cc:4000
0x67388d cgraph_node::expand()
        /home/ansuel/openwrt/build_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/gcc-12.2.0/gcc/cgraphunit.cc:1806
0x67388d cgraph_node::expand()
        /home/ansuel/openwrt/build_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/gcc-12.2.0/gcc/cgraphunit.cc:1788
0x674ea6 expand_all_functions
        /home/ansuel/openwrt/build_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/gcc-12.2.0/gcc/cgraphunit.cc:1999
0x674ea6 symbol_table::compile()
        /home/ansuel/openwrt/build_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/gcc-12.2.0/gcc/cgraphunit.cc:2349
0x5e59b1 lto_main()
        /home/ansuel/openwrt/build_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/gcc-12.2.0/gcc/lto/lto.cc:657
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <http://bugs.openwrt.org/> for instructions.
lto-wrapper: fatal error: /home/ansuel/openwrt/staging_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/bin/mipsel-openwrt-linux-musl-gcc returned 1 exit status
compilation terminated.
/home/ansuel/openwrt/staging_dir/toolchain-mipsel_24kc_gcc-12.2.0_musl/lib/gcc/mipsel-openwrt-linux-musl/12.2.0/../../../../mipsel-openwrt-linux-musl/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
make[7]: *** [Makefile:566: libnftables.la] Error 1
make[7]: Leaving directory '/home/ansuel/openwrt/build_dir/target-mipsel_24kc_musl/nftables-json/nftables-1.0.5/src'
make[6]: *** [Makefile:481: all] Error 2
make[6]: Leaving directory '/home/ansuel/openwrt/build_dir/target-mipsel_24kc_musl/nftables-json/nftables-1.0.5/src'
make[5]: *** [Makefile:478: all-recursive] Error 1
make[5]: Leaving directory '/home/ansuel/openwrt/build_dir/target-mipsel_24kc_musl/nftables-json/nftables-1.0.5'
make[4]: *** [Makefile:387: all] Error 2
make[4]: Leaving directory '/home/ansuel/openwrt/build_dir/target-mipsel_24kc_musl/nftables-json/nftables-1.0.5'
make[3]: *** [Makefile:88: /home/ansuel/openwrt/build_dir/target-mipsel_24kc_musl/nftables-json/nftables-1.0.5/.built] Error 2
make[3]: Leaving directory '/home/ansuel/openwrt/package/network/utils/nftables'
time: package/network/utils/nftables/json/compile#9.29#0.83#10.84
    ERROR: package/network/utils/nftables failed to build (build variant: json).
make[2]: *** [package/Makefile:118: package/network/utils/nftables/compile] Error 1
make[2]: Leaving directory '/home/ansuel/openwrt'
make[1]: *** [package/Makefile:112: /home/ansuel/openwrt/staging_dir/target-mipsel_24kc_musl/stamp/.package_compile] Error 2
make[1]: Leaving directory '/home/ansuel/openwrt'
make: *** [/home/ansuel/openwrt/include/toplevel.mk:231: world] Error 2
```

is this the same?

neheb commented:
Yes

mpratt14 commented:
@Ansuel is it still "only on ubuntu"?

Ansuel commented:
@mpratt14 no idea today i was rebuiliding an image for mipsel and i still had toolchain for gcc 12.1 so they got recompiled and i have this... this is on ubuntu devel

neheb commented:
I've only seen this on Ubuntu. Debian and Fedora works fine. No idea why.

jmccrohan commented:
> I've only seen this on Ubuntu. Debian and Fedora works fine. No idea why.

What version of Debian did you test with? I can confirm same behaviour on Debian testing/bookworm.

neheb commented:
Lightweight nspawn Debian 10 container.

dhewg committed:
toolchain/gcc: clean up CFLAGS

Instead of having two different ways to pass flags to the gcc build
process, add them as configure args, which is a reliable way to let
gcc pass them around to its various pieces.

Also add CXXFLAGS, since gcc started to use c++ for itself recently
(~10 years ago now).

Signed-off-by: Andre Heider <a.heider@gmail.com>

dhewg committed:
toolchain/binutils: use explicit configure args

Spell out what we want to enable or disable. This prevents host libs to leak in,
so everyone get the same feature set.

Signed-off-by: Andre Heider <a.heider@gmail.com>

dhewg committed:
toolchain/gcc: use explicit configure args

Spell out what we want to enable or disable. This prevents host libs to leak in,
so everyone get the same feature set.

Signed-off-by: Andre Heider <a.heider@gmail.com>

neheb commented:
@anomeome rebased with some extra PRs. Can you retest?

anomeome commented:
Successfully built with _r21992-19988b66d0_, have not flashed an image.

mpratt14 commented:
@neheb recall that @anomeome was able to build successfully before these new commits on a brand new buildroot so basically, you're the only one out of the 4 of us who can reproduce the original problem

neheb commented:
I see @Ansuel posted a log of the problem. PR currently should work as it disables zlib in toolchain. zstd can be reenabled in a different PR and retested.

dhewg commented:
`--with-system-zlib` is supposed to pick up tools/zlib, since the gcc build process is fed with $HOST_CFLAGS, which includes the according -I/-L flags.

That works here and on CI, but it needs at least some more testing to make sure that won't regress. Especially on macos, I can only guess if those commits break it, but I hope they won't ;)
But the idea is the same: we build it, use it, so everyone gets the same result.

The alternative is the remove that switch again, and that makes gcc use its bundled zlib instead.

zstd is handled differently. It isn't bundled as zlib is. You need to provide it if you want that feature.

That's why there's no problem with a host zlib leaking in, but a zstd one is.

dhewg commented:
We can also consider this patch to detect these kind of issues
https://github.com/LibreELEC/LibreELEC.tv/blob/master/packages/lang/gcc/patches/gcc-crosscompile-badness.patch

I didn't check if that works on gcc itself, but with it's multiple phases to build a cross compiler it even might

neheb commented:
That looks like a very good patch to try with CONFIG_ALL.

neheb commented:
removed draft

@Ansuel this is ready.

dhewg commented:
Can we drop the gcc patch, and add that later with another PR enabling zstd? Sounds like it's only required for that, any maybe there's another patch-less way to solve that (as this also triggers autoreconf if I had to guess).

Does this PR now fix the affected systems that were mentioned in here? Did someone build test on macos?

dhewg commented:
Follow-up on zstd, which isn't enabled with the patches here (which is good for now!):
I just realized that --with-zstd is new with binutils 2.40, which is what I use, but it's not even the default yet. I also didn't check gcc 11 about zstd, which is still supported.

There seems to be some interplay between the two packages w.r.t. zstd. I'm not sure what versions you can mix when compressing debug sections and/or LTO object files with zstd. It may be that you need both, gcc 12 and binutils 2.40 for this.

Anyway, with your patches to build only a static zstd, its .pc already has `-pthread` in c/lflags. binutils uses pkg-config, so switching to `--with-zstd` works there and it always links with `-lzstd -pthread`. So far so good.

gcc doesn't use pkg-config though. Squeezing in a `-pthread` works via `ac_cv_search_ZSTD_compress=-lzstd -pthread`:
`checking for library containing ZSTD_compress... (cached) -lzstd -pthread`

That works for me, see https://github.com/dhewg/openwrt/commits/toolchain

neheb commented:
@dhewg removed.

neheb commented:
@dhewg

```
/usr/bin/ld: cannot find -lz
```

not sure if related to this PR. Bad that it's leaking from the host.

dhewg commented:
Huh, I looked at the configure scripts, but it didn't look like --with-system-zlib=path was supported. Maybe in a subdir, that auto* stuff on gcc can get confusing with all those pieces. You basically have to check */configure and can pass args even if the root script doesn't even mention it.

But glad that fixed it, feel free to just fixup my patch there that left it out.
Keep me posted if you get any macos reports

neheb commented:
Fixed and readded zstd commits.

I did a run with CONFIG_ALL. Everything builds (just base feed).

dhewg commented:
That's the easy one, I use those patches locally too. But does it work with gcc11 and/or older binutils versions?

dhewg commented:
Here's an article which tracks zstd progress: https://maskray.me/blog/2022-09-09-zstd-compressed-debug-sections

Some detective work:

binutils master vs v2.40 vs v2.39:
https://sourceware.org/git/?p=binutils-gdb.git&a=search&h=HEAD&st=commit&s=zstd
https://sourceware.org/git/?p=binutils-gdb.git&a=search&h=refs%2Fheads%2Fbinutils-2_40-branch&st=commit&s=zstd
https://sourceware.org/git/?p=binutils-gdb.git&a=search&h=refs%2Fheads%2Fbinutils-2_39-branch&st=commit&s=zstd
-> v2.40 matches master, zstd non-existent in v2.39

gcc master vs v12.2.0 vs v11.3.0:
https://gcc.gnu.org/git/?p=gcc.git&a=search&h=HEAD&st=commit&s=zstd
https://gcc.gnu.org/git/?p=gcc.git&a=search&h=refs%2Ftags%2Freleases%2Fgcc-12.2.0&st=commit&s=zstd
https://gcc.gnu.org/git/?p=gcc.git&a=search&h=refs%2Ftags%2Freleases%2Fgcc-11.3.0&st=commit&s=zstd
-> master gained  -gz=zstd and zstd support for libbacktrace

elfutils master vs v0.188:
https://sourceware.org/git/?p=elfutils.git&a=search&h=HEAD&st=commit&s=zstd
https://sourceware.org/git/?p=elfutils.git&a=search&h=refs%2Ftags%2Felfutils-0.188&st=commit&s=zstd
-> master gained zstd support

Based on those git logs:
Enabling zstd here should at least check binutils >= 2.40, the supported gcc versions don't appear to matter.
I'm not sure how relevant libbacktrace and libelf are in the zstd context.

neheb commented:
My compilation test was with defaults, so binutils 2.39 and Gucci 12.2

dhewg commented:
Okay, did you somehow verify if LTO enabled packages already use zstd with that?

binutils ignores unknown configure switches, so --with-zstd old <v2.40
Maybe that's all fine for LTO as is.

I was looking into compressed debug sections. That's where all the above tools come in. I verified that it's working with zlib, for zstd we need to wait for gcc13's --with-gz=zstd anyway

So assuming LTO is okay we can leave it as-is? (But I think you shuffled the commits, tools/cmake should be after tools/zstd?)

neheb commented:
rebased

dhewg commented:
Build tested with binutils 2.37 and gcc 11 for a ramips target, no issues apart from an unrelated package/elfutils error

dhewg commented:
Build tested with binutils 2.38 and gcc 12 for a lantiq target

dhewg commented:
Build and runtime tested with binutils 2.40 and gcc 11 for armvirt64, no issues either

Looks good to me, but we really could use a macos test drive

anomeome commented:
[failed](https://termbin.com/w7kw) for me with target mvebu wrtpac devices, GCC12 and binutils 2.40

neheb commented:
:( looks like I have to get rid of the zstd stuff again

anomeome commented:
Started from a

`rm -rf bin tmp build_dir staging_dir`

dhewg commented:
better wipe everything: https://www.mail-archive.com/openembedded-core@lists.openembedded.org/msg158289.html
apparently that error can happen on prior non-zstd lto object files

anomeome commented:
`make dirclean` yielded same result.

neheb commented:
@anomeome how about now?

anomeome commented:
Successful build, no flash due to the outstanding kernel boot issue.

dhewg commented:
@anomeome
With the zstd patches still in place, can you try binutils 2.39 with gcc 12 please, so we know which part messes things up?
If you have a toolchain that fails with build ubus, could you then please tarxz and upload that somewhere, so we can inspect it?

dhewg commented:
Oh, and a build logs too please ;)

```
make dirclean
make -j8 tools/compile
make V=cs -j8 toolchain/compile 2>&1|tee /tmp/toolchain.log
make V=cs -j8 package/ubus/compile 2>&1|tee /tmp/ubus.log
```

dhewg commented:
A debootstrapped jammy running under lxc builds everything just fine here, not matter if the older libzstd1 v1.4.8 or even libzstd-dev is installed.

@anomeome I used this branch with the default binutils-2.39 and gcc-12.2.0, can you give that a try? I can only think of you accidentally using outdated patches, I recommend you fetch that directly:
https://github.com/dhewg/openwrt/commits/toolchain

mpratt14 commented on toolchain/binutils/Makefile:
```diff
@@ -43,8 +43,12 @@ HOST_CONFIGURE_ARGS = \
 	--host=$(GNU_HOST_NAME) \
 	--target=$(REAL_GNU_TARGET_NAME) \
 	--with-sysroot=$(TOOLCHAIN_DIR) \
+	--with-system-zlib \
```
are we not able to specify a custom directory here like with gcc?

dhewg replied:
That's just a bool flag AFAICS:
```
bfd/configure:with_system_zlib
bfd/configure:if test "${with_system_zlib+set}" = set; then :
bfd/configure:  withval=$with_system_zlib; if test x$with_system_zlib = xyes ; then
binutils/configure:with_system_zlib
binutils/configure:if test "${with_system_zlib+set}" = set; then :
binutils/configure:  withval=$with_system_zlib; if test x$with_system_zlib = xyes ; then
config/zlib.m4:  if test x$with_system_zlib = xyes ; then
configure:with_system_zlib
configure:if test "${with_system_zlib+set}" = set; then :
configure:  withval=$with_system_zlib;
configure:if test x$with_system_zlib = xyes ; then
configure:# Set with_gnu_as, with_gnu_ld, and with_system_zlib as appropriate.
configure:  with_system_zlib=yes
configure.ac:if test x$with_system_zlib = xyes ; then
configure.ac:# Set with_gnu_as, with_gnu_ld, and with_system_zlib as appropriate.
configure.ac:  with_system_zlib=yes
gas/configure:with_system_zlib
gas/configure:if test "${with_system_zlib+set}" = set; then :
gas/configure:  withval=$with_system_zlib; if test x$with_system_zlib = xyes ; then
gdb/configure:with_system_zlib
gdb/configure:if test "${with_system_zlib+set}" = set; then :
gdb/configure:  withval=$with_system_zlib; if test x$with_system_zlib = xyes ; then
gold/configure:with_system_zlib
gold/configure:if test "${with_system_zlib+set}" = set; then :
gold/configure:  withval=$with_system_zlib; if test x$with_system_zlib = xyes ; then
gprofng/configure:with_system_zlib
gprofng/configure:if test "${with_system_zlib+set}" = set; then :
gprofng/configure:  withval=$with_system_zlib; if test x$with_system_zlib = xyes ; then
ld/configure:with_system_zlib
ld/configure:if test "${with_system_zlib+set}" = set; then :
ld/configure:  withval=$with_system_zlib; if test x$with_system_zlib = xyes ; then
libctf/configure:with_system_zlib
libctf/configure:if test "${with_system_zlib+set}" = set; then :
libctf/configure:  withval=$with_system_zlib; if test x$with_system_zlib = xyes ; then
zlib/configure:with_system_zlib
zlib/configure:if test "${with_system_zlib+set}" = set; then :
zlib/configure:  withval=$with_system_zlib;
zlib/configure:   if test "$with_system_zlib" = yes; then
zlib/configure.ac:   if test "$with_system_zlib" = yes; then
```

But as I mentioned before, it gets HOST_*FLAGS to find our tools/ version.

zstd is found via pkg-config, which works because it includes toolchain-build.mk which includes host-build.mk which sets PKG_CONFIG_{PATH,LIB}:
```
bfd/configure:  --with-zstd             support zstd compressed debug sections
bfd/configure:# Check whether --with-zstd was given.
bfd/configure:      as_fn_error $? "--with-zstd was given, but pkgconfig/libzstd.pc is not found" "$LINENO" 5
bfd/configure:      as_fn_error $? "--with-zstd was given, but pkgconfig/libzstd.pc is not found" "$LINENO" 5
binutils/configure:  --with-zstd             support zstd compressed debug sections
binutils/configure:# Check whether --with-zstd was given.
binutils/configure:      as_fn_error $? "--with-zstd was given, but pkgconfig/libzstd.pc is not found" "$LINENO" 5
binutils/configure:      as_fn_error $? "--with-zstd was given, but pkgconfig/libzstd.pc is not found" "$LINENO" 5
config/zstd.m4:  [AS_HELP_STRING([--with-zstd], [support zstd compressed debug sections (default=auto)])],
config/zstd.m4:      AC_MSG_ERROR([--with-zstd was given, but pkgconfig/libzstd.pc is not found])
configure:  --with-zstd             Support zstd compressed debug sections
configure:# Check whether --with-zstd was given.
configure.ac:[AS_HELP_STRING([--with-zstd], [Support zstd compressed debug sections (default=auto)])])
gas/configure:  --with-zstd             support zstd compressed debug sections
gas/configure:# Check whether --with-zstd was given.
gas/configure:      as_fn_error $? "--with-zstd was given, but pkgconfig/libzstd.pc is not found" "$LINENO" 5
gas/configure:      as_fn_error $? "--with-zstd was given, but pkgconfig/libzstd.pc is not found" "$LINENO" 5
gdb/configure:  --with-zstd             support zstd compressed debug sections
gdb/configure:# Check whether --with-zstd was given.
gdb/configure:      as_fn_error $? "--with-zstd was given, but pkgconfig/libzstd.pc is not found" "$LINENO" 5
gdb/configure:      as_fn_error $? "--with-zstd was given, but pkgconfig/libzstd.pc is not found" "$LINENO" 5
gold/configure:  --with-zstd             support zstd compressed debug sections
gold/configure:# Check whether --with-zstd was given.
gold/configure:      as_fn_error $? "--with-zstd was given, but pkgconfig/libzstd.pc is not found" "$LINENO" 5
gold/configure:      as_fn_error $? "--with-zstd was given, but pkgconfig/libzstd.pc is not found" "$LINENO" 5
ld/configure:  --with-zstd             support zstd compressed debug sections
ld/configure:# Check whether --with-zstd was given.
ld/configure:      as_fn_error $? "--with-zstd was given, but pkgconfig/libzstd.pc is not found" "$LINENO" 5
ld/configure:      as_fn_error $? "--with-zstd was given, but pkgconfig/libzstd.pc is not found" "$LINENO" 5
```

But since it's such a PITA to build gcc with zstd let's leave that for later...

anomeome commented:
@dhewg I'm a bit confused, my last post was that with the latest push of PR things built successfully. As far as outdated patches, very doubtful
```
git reset --hard
git status
```
followed by current version of desired patches. At any rate things also built successfully changing to binutils 2.39.

@neheb fix for kernel boot issue is in, image built with PR flashed and running on _mamba_.

dhewg commented:
neheb removed the zstd patches, that's why it builds now ;)

anomeome commented:
So you are asking me to backup to the none working version?

dhewg commented:
Yes! ;)
Well in the long run we want zstd support, but we need to figure out why it fails to build for some. I tried pretty hard, but I fail to break it.
But hang on with testing, I think I may have found a possible cause

dhewg commented:
Okay, please take this branch for a test drive, I just cannot reproduce the build failure you're getting when zstd patches are included:
https://github.com/dhewg/openwrt/commits/toolchain

anomeome commented:
OK, but I cannot even apply [this](https://github.com/dhewg/openwrt/commit/e12d688de2471043b6c3943bb3dad7146deb58d4), fails as the file is different (even on a fresh clone), your tree dated perhaps.

dhewg commented:
Hm, I rebased and pushed just before my last comment.

In your OpenWrt git clone:
```
git remote add dhewg https://github.com/dhewg/openwrt.git
git fetch dhewg
git checkout -b toolchain dhewg/toolchain
```
And then the whole clean/build dance.

That gives just exactly my branch without fetching&applying patches (which is what I meant before. You could also do a fresh clone, but that's working against git, not with ;)

anomeome commented:
Right, same error as neheb's last zstd push as previously posted, [log file](https://termbin.com/eq19) from:

`make V=sc -j5 package/ubus/{clean,compile} 2>&1|tee /tmp/ubus.log`

dhewg commented:
Sigh, oh well, thanks for the test!
Can you post your .config? The termbin link above is expired

anomeome commented:
[configdiif](https://termbin.com/1qq6s)

neheb commented:
The problem is GCC picking something up from the host. I don't know what. Maybe I need to set up a VM with snapshots to figure out what.

anomeome commented:
But apparently  what just failed for me on bare metal was successful in a LXC?

dhewg commented:
Yeah, minimal jammy LXC rootfs works here, without host libzstd, with it, and with it plus its -devel package.
Maybe some other package, some combination or something completely different and silly.

anomeome commented:
Another data point, might just discombobulate the confusion, but I did:
```
 2179  git clone https://git.openwrt.org/openwrt/openwrt.git owrt
 2180  cd owrt/
 2181  ls -las
 2182  git remote add dhewg https://github.com/dhewg/openwrt.git
 2183  git fetch dhewg
 2184  git checkout -b toolchain dhewg/toolchain
 2185  ./scripts/feeds update -a
 2186  ./scripts/feeds install -a
 2187  cp ../source/configdiff2 .config
 2188  make defconfig && make oldconfig
```
and built to success. So new buildroot and packages. no old _/dl_ directory contents...

Is it possible that there is some _dried muffin remnant_ somewhere that is not detected by git, nor cleaned out by dirclean, yet the make system finds and uses same.

Ansuel commented:
@dhewg any suggestion on how to setup this? also I'm curious if we should add this kind of thing for our CI test

neheb commented:
Can we get this merged? I or @dhewg can post a new PR adding zstd support to GCC. We can continue the discussion there.

Ansuel commented:
@neheb do we need to update the pr title / description since something has changed?

neheb commented:
done.

Ansuel commented:
And just to be sure using zlib instead of wrongly using host zstd just cause bigger toolchain right?

neheb commented:
it shouldn't. Before this PR, GCC picks up host zlib if available (in most cases, yes). Now, it uses tools/zlib.

Ansuel commented:
And use zstd if present in the system right?

neheb commented:
Yeah. Before this PR, GCC picks up host zlib if available (in most cases, no). Now, it's explicitly disabled as tools/zstd works but GCC picks up something else from the host and breaks.

Ansuel commented:
@neheb hope the use of zstd won't be impossible

neheb commented:
it will be once the offending package is found.

my current idea is to use snapshots to install various packages and see when it breaks.

dhewg commented:
> and built to success.

Wait, what, I works now? I force pushed yet another attempt yesterday, what's the top hash of `git log`?

It is possible that some old crap was still lying around that's responsible for all this. You could try `git clean -fdxn` (`n` only prints what it would wipe, remove to actually wipe, but note that includes e.g.your .config too)

dhewg commented:
> any suggestion on how to setup this? also I'm curious if we should add this kind of thing for our CI test

I tried hard to reproduce, but failed to do so. With the latest comments it may be old .o files that won't get wiped, which then are not rebuilt and gcc ice'ing because it expects zstd compressed lto objects which were zlib compressed from an earlier toolchain. Or something. We're down to a guessing game here :\

anomeome commented:
Ya, but to a new clone of a buildroot, same thing failed in my existing buildroot. After that post, it occurred to me that you may have made a change, so I checked your repo and there was one timestamp updated, but the diff looked to be the same (at least to my memory bank). I had already deleted the tree at that point.

I did try a build with new neheb PR yesterday on my existing buildroot, ii built to the same error, good thing in a backhanded kind of way I suppose.

This begins to feel like an incomplete implementation of the semantics expected by a _make dirclean_. I got caught out by that a few years back, IIRC some generated .py files in /config.

dhewg commented:
What we usually do is amend patches, so even if they look the same at first glance they're different then.
To add to the confusion, what I pushed yesterday is newer than what @neheb pushed to the new PR - even though the two patches have the same title.

If I understood correctly, #11970 is what failed again? How about my current toolchain branch? That's now rebased on top of other merged toolchain patches (so just the two zstd ones on top of master)

Assuming you have my remote and branch still is your clone:
```
git fetch dhewg
git checkout toolchain
git reset --hard dhewg/toolchain
```

The amending and rebasing can get confusing, but at the same time it's the weapon of choice to incorporate change requests for github PRs.

FWIW I tried a dirclean too, and it looked like everything got wiped correctly. But if one clone dir works and not the other there has to be some difference?

anomeome commented:
The above in my original (guaranteed to fail) buildroot, from a _make dirclean_ same error. Try again from:

```
make distclean
./scripts/feeds update -a
./scripts/feeds install -a
cp configdiff2 .config
make defconfig && make oldconfig
```
builds to success.

So it would seem there is a delta from _distclean_ to _dirclean_ that is removing the culprit. And in hindsight I really should have captured a _ls -lasR_ post a `rm -rf bin tmp build_dir staging_dir` to attempt to ID same.

dhewg commented:
distclean wipes .ccache, and your config has CONFIG_CCACHE=y
So I bet it's just that crap. I stopped using it ~10 years ago because it wastes your time with these kind of errors.
I consider the patches working now :P

@neheb could you please update the other PR with my lastest patches? https://github.com/dhewg/openwrt/commits/toolchain

neheb commented:
that is...very interesting

I have ccache disabled as it doesn't build on Ubuntu 18.04. Someone needs to add a PR for that :)

dhewg commented:
Yeah, I didn't like ccache before, now I hate it :D
I left a note in the gcc zstd commit message.

We (by which I mean not me :P) could add a patch that wipes the ccache cache for the dirclean target. Because it's not just this lto/zstd thing, toolchain changes might have no effect, as ccache fails it and might return object files from an earlier toolchain version /o\

dhewg commented:
dirclean also hardcodes .ccache and doesn't respect CONFIG_CCACHE_DIR, it doesn't get any better, does it?

Since CI and buildbot have some ccache grep hits too, cc @Ansuel for awareness before considering #11970

Ansuel commented:
@dhewg this really seems to be a bug with ccache or for sure a corner case... and looks interesting that this was the problem all along... 

Would be good to fix the bug of dirclean not following CONFIG_CCACHE_DIR option

dhewg commented:
Depends on who you ask, I guess. It's a bug, but it's also a fundamental problem with ccache and always will be.

I don't know what its constraints are to trigger a rebuild instead of using the cached version, but taking the wrong decision there is the root of all ccache related heisenbugs. lto/zstd is a first, but the ones I've dealt with are, boiled down, in exactly that category.

Anyway, not a fan as you may have noticed, I just wanted to make sure that buildbot/CI are prepared before adding zstd support.

Ansuel commented:
@dhewg an idea may be to wipe ccache on toolchain and/or host changes... That should solve any problem

dhewg commented:
Yeah. ccache detects some changes at least. I guess a wipe upon toolchain/ changes would be enough

Ansuel commented:
But IMHO we should first alert ccache project that this may happen... this new discovery that ccache is the culprit should give us a way to correctly create a repro env...

neheb commented:
That dirclean patch should be written, yes

dhewg commented:
@anomeome could you reproduce and verify if `rm -rf .ccache` fixes the build error?

anomeome commented:
Much to my surprise I was able to get back to that special place where things break, the [output of:](https://termbin.com/6omn)

```
nice -n -1 ionice -c2 -n0 make --jobs=5
make V=sc -j5 package/ubus/{clean,compile}
rm -rf .ccache/*
make V=sc -j5 package/ubus/{clean,compile}
```

neheb commented:
wonder if it's being miscompiled by GCC9. I don't see an issue with GCC12.

dhewg commented:
ccache will probably detect a change on the gcc version, at least I hope so.

Here we're a bit mean to ccache. There's a gcc 12.2.0 without zstd support, so lto object files are zlib compressed.
Then we're not bumping any toolchain component, but just enabling zstd support.

So the toolchain still looks the same to ccache, and when it gets to compiling ubus with lto, it returns the object files from an earlier toolchain, so zlib and not zstd compressed hence gcc ice'ing.

It could detect the toolchain change with timestamps of the $CC binary, but obviously it doesn't do that.

The last log confirms that, so we *need* to clear the ccache to prevent this issue popping up on buildbot/CI (if they're using ccache)

mpratt14 commented:
> It could detect the toolchain change with timestamps of the $CC binary, but obviously it doesn't do that.
> 
> The last log confirms that, so we *need* to clear the ccache to prevent this issue popping up on buildbot/CI (if they're using ccache)

the build system is set up to assume that if you change something in toolchain, for example, that it would be followed by cleaning the entire buildroot with dirclean or by running make world which will cause a rebuild there anyway. running make on an individual package will not ensure that a change somewhere else will result in a rebuild of that (in other words, individual build tools are not explicitly prereqs of individual build targets)

instead of timestamps of the binary, it would be a list of things in source that altogether affect the resulting binary, and you can add files to that list 

perhaps there could be a target within prereq target to clear ccache under certain conditions, but for now its the responsibility of the user to clean the buildroot, and it's the build system's responsibility to keep up to date only the targets that the user specifies

maybe something like, let toolchain/clean have a line for removing ccache data, and then make toolchain/clean a prereq of toolchain/install

mpratt14 commented:
in terms of automatically cleaning and rebuilding the toolchain, we should look at db34b93331e91bdb2cbc15d17632aaaab7217a6d

maybe its causing problems or not doing the job and can be improved
